shared/trie/deposit_trie.go

```diff
@@ -0,0 +1,46 @@
+package trie
+
+import (
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// DepositTrie represents a Merkle trie tracking deposits on the ETH 1.0
+// PoW chain contract created in Vyper.
+type DepositTrie struct {
+	depositCount uint64
+	merkleHashes map[uint64][32]byte
+}
+
+// NewDepositTrie creates a new struct instance with a hash list of initial
+// length equal to 2 to the power of the deposit contract's tree depth.
+func NewDepositTrie() *DepositTrie {
+	return &DepositTrie{
+		depositCount: 0,
+		merkleHashes: make(map[uint64][32]byte),
+	}
+}
+
+// UpdateDepositTrie updates the Merkle trie representing deposits on
+// the ETH 1.0 PoW chain contract.
+func (d *DepositTrie) UpdateDepositTrie(depositData []byte) {
+	twoToPowerOfTreeDepth := 1 << params.BeaconConfig().DepositContractTreeDepth
+	index := d.depositCount + uint64(twoToPowerOfTreeDepth)
+	d.merkleHashes[index] = hashutil.Hash(depositData)
+	for i := uint64(0); i < params.BeaconConfig().DepositContractTreeDepth; i++ {
+		index = index / 2
+		left := d.merkleHashes[index*2]
+		right := d.merkleHashes[index*2+1]
+		if right == [32]byte{} {
```

Comment:
The default leaves are of the form 0x00000000000000000000000000000, so if we have a left leaf of 0x213019239012031239123 and a right leaf of 0x0000000000000000000 we do not want their parent to be hash(0x2130192390120312391230000000000000000000), but rather just hash(0x213019239012031239123)